coralproject/talk
# [CORL-2799]: Single comment embed

## What does this PR do?

This PR adds support for a single comment embed of a Coral comment.

It includes the ability to use the `Embed comment` link in moderation actions for a comment in the stream to copy a simple comment embed to paste in a story. Then a jsonp script will replace the fallback comment with the full comment embed.

It also supports using the Coral Oembed provider API to embed comments in stories.

- [x] Return proper Oembed errors, including not implemented for XML Oembed request and not authorized for non-whitelisted URL
- [x] Update permalink to support embed links
- [x] Add button to click to get comment embed code
- [x] Add `embeddedAt` timestamp to comments to track when they are embedded
- [x] Update documentation regarding comment embed class names and styles, and how it works
- [x] Handle spoiler and sarcasm within comment body and sanitization of comment body in embeds
- [x] Add admin configuration for whether interactions are included in comment embeds
- [x] Add/update tests related to the single comment embed
- [x] Support localization in comment embeds

## These changes will impact:


## What changes to the GraphQL/Database Schema does this PR introduce?



## Does this PR introduce any new environment variables or feature flags?

no

## If any indexes were added, were they added to `INDEXES.md`?

none

## How do I test this PR?

You can update `commentEmbed.html` to test out comment embeds. Find the comment that you want to embed from the stream, open the moderation actions, and click `Embed comment`. This will copy the embed code to your clipboard; paste into `commentEmbed.html` where you want to see the embedded comment. Then you can run Coral and navigate to `/commentEmbed.html` and see your comments replaced with comment embed shadow doms. 

To test out the fallback comments, remove the `coral-comment-embed` class from the embed code and refresh `commentEmbed.html`.

You can add custom fonts and styles in the admin in advanced configuration, and these will then be reflected in your embeds, too.

Test out embedded Tweets, Youtube, external media, and gifs. See that they look as expected.

Test out comments with all formatting, including spoilers, sarcasm, and blockquotes, within the full comment embeds and the fallback comments.

Test out changing the embedded comments allow replies configuration under advanced in the admin and seeing that change reflected in the embedded comments. See if you update and add a `data-allowReplies` attribute and set to `"true"` or `"false"`, this will override the setting from the admin.

See that clicking on the embedded comment takes you to its permalink page. Same with clicking on `Reply`, `Rec` or `Go to conversation`. See that each has its own query parameter for event tracking.

Check comments that are embedded have an `embeddedAt` timestamp added.

Test out localization in embedded comments. Change your locale, and see that button text is updated.

Test it only works on whitelisted urls.

## Where any tests migrated to React Testing Library?

Yes, `src/core/client/admin/test/configure/advanced.spec.tsx`

## How do we deploy this PR?
